src/Analyzers/MSTest.Analyzers/DataRowShouldBeValidAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+        foreach (var attribute in dataRowAttributes)
+        {
+            AnalyzeAttribute(context, attribute, methodSymbol);
+        }
+    }
+
+    private static void AnalyzeAttribute(SymbolAnalysisContext context, AttributeData attribute, IMethodSymbol methodSymbol)
+    {
+        if (attribute.ApplicationSyntaxReference?.GetSyntax() is not { } syntax)
+        {
+            return;
+        }
+
+        // No constructor arguments and no method parameters -> nothing to check.
+        if (attribute.ConstructorArguments.Length == 0 && methodSymbol.Parameters.Length == 0)
+        {
+            return;
+        }
+
+        // Count mismatch since there's zero method parameters but there's at least one DataRow
+        // constructor argument.
+        if (methodSymbol.Parameters.Length == 0)
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                attribute.ConstructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Possible count mismatch depending on whether last method parameter is an array or not.
+        IParameterSymbol lastMethodParameter = methodSymbol.Parameters.Last();
+        bool lastMethodParameterIsArray = lastMethodParameter.Type.Kind == SymbolKind.ArrayType;
+        if (attribute.ConstructorArguments.Length == 0)
+        {
+            if (!lastMethodParameterIsArray)
+            {
+                context.ReportDiagnostic(syntax.CreateDiagnostic(
+                    ArgumentCountMismatchRule,
+                    attribute.ConstructorArguments.Length,
+                    methodSymbol.Parameters.Length));
+            }
+
+            return;
+        }
+
+        // DataRow constructors have either zero or one argument(s). If we get here, we are
+        // on the one argument case. Check if we match either of the array argument constructors
+        // and expand the array argument if we do.
+        ImmutableArray<TypedConstant> constructorArguments = attribute.ConstructorArguments;
+        if (constructorArguments[0].Kind is TypedConstantKind.Array && !constructorArguments[0].IsNull)
+        {
+            constructorArguments = constructorArguments[0].Values;
+        }
+
+        if (IsArgumentCountMismatch(constructorArguments.Length, methodSymbol.Parameters.Length, lastMethodParameterIsArray))
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                constructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Check constructor argument types match method parameter types.
+        List<(int ConstructorArgumentIndex, int MethodParameterIndex)> typeMismatchIndices = new();
+        for (int i = 0, j = 0; i < constructorArguments.Length; ++i, ++j)
+        {
+            if (constructorArguments[i].IsNull)
```

Comment:
I'll add a comment here in code to say that `null` is considered as `default` for non-nullable types.

Reply:
Fixed, thanks.